tchannel-core/src/main/java/com/uber/tchannel/api/TFuture.java

```diff
@@ -1,26 +1,121 @@
+/*
+ * Copyright (c) 2015 Uber Technologies, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
 package com.uber.tchannel.api;
 
 import com.google.common.util.concurrent.AbstractFuture;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.uber.tchannel.api.handlers.TFutureCallback;
+import com.uber.tchannel.errors.ErrorType;
+import com.uber.tchannel.headers.ArgScheme;
+import com.uber.tchannel.messages.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+public final class TFuture<V extends Response> extends AbstractFuture<V> {
+
+    private static final Logger logger = LoggerFactory.getLogger(TFuture.class);
+
+    @SuppressWarnings({"rawtypes"})
+    public static TFuture create(ArgScheme argScheme) {
+        return new TFuture(argScheme);
+    }
+
+    private AtomicInteger listenerCount = new AtomicInteger(0);
+    private V response = null;
+    private ArgScheme argScheme = null;
 
-public final class TFuture<V> extends AbstractFuture<V> {
-    public static <V> TFuture<V> create() {
-        return new TFuture();
+    private TFuture(ArgScheme argScheme) {
+        this.argScheme = argScheme;
     }
 
-    private TFuture() {
+    @SuppressWarnings({"unchecked"})
+    public void addCallback(final TFutureCallback<V> callback) {
```

Comment:
The reason I added this interface is because the original interface support Futures.addCallback: 
e.g., https://github.com/uber/tchannel-java/blob/master/tchannel-hyperbahn/src/main/java/com/uber/tchannel/hyperbahn/api/HyperbahnClient.java#L102

I want to deprecate that. 

BTW, why addListener is better? The callback is implemented based on addListener with better error handlings.